in zabbix_host check for empty proxy in existing host entry, fixes #24407

mschuett opened the pull request:
##### SUMMARY

If the host exists then `proxy_id` is always set to None and will never match the data from host['proxy_hostid'], hence an additional check is necessary.

Bug and fix were described in https://github.com/ansible/ansible-modules-extras/issues/3280, it only got lost at some time in between release 2.2 and 2.3.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
module zabbix_host

##### ANSIBLE VERSION
```
ansible 2.3.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.5 (default, Aug  2 2016, 04:20:16) [GCC 4.8.5 20150623 (Red Hat 4.8.5-4)]
```

mschuett committed:
fixes #24407

angystardust commented:
Sorry to say that but this patch still doesn't fix the idempotency of the zabbix_host module.
If I launch 2 times the following task, I always get a "changed" state even if the zabbix host was already present.

```
- name: Create a new host or update an existing host's info
# local_action:
#   module: zabbix_host
  local_action:
    module: zabbix_host
    server_url: https://test.zabbix.internal/
    login_user: admin
    login_password: password
    host_name: 'test-ansible'
    host_groups:
      - "Zabbix servers"
    link_templates:
      - Template OS Linux
    status: disabled
    state: present
```

mschuett commented:
Right now I can only share the observation "works for me".
If time permits I will try to investigate further and replicate the setup in Vagrant.

angystardust commented:
Thanks for your time, @mschuett. FYI, I tested the module wth both zabbix 2.4 and zabbix 3.2 but I'm having the same issue.

calfonso commented:
@mschuett have time to address the failed test result from @angystardust ?

mschuett commented:
I finally got around to write a small test setup in https://github.com/mschuett/test_ansible_pr_24407
It starts two virtual machines (with CentOS 7.3 and Ansible 2.3.2.0) and installs Zabbix 3.2.
Using the library subfolder one can test different zabbix_host module versions.

When I run this on my machine (`vagrant up`) then I observer the same pattern as on my servers: the PR version is idempotent, the one from the current devel branch is not:
```
TASK [add with zabbix_host_pr_24407] *******************************************
changed: [zabbix -> localhost]
changed: [workstation -> localhost]

TASK [add with zabbix_host_pr_24407] *******************************************
ok: [zabbix -> localhost]
ok: [workstation -> localhost]

TASK [add with zabbix_host_ansible_devel] **************************************
changed: [zabbix -> localhost]
changed: [workstation -> localhost]

TASK [add with zabbix_host_ansible_devel] **************************************
changed: [workstation -> localhost]
changed: [zabbix -> localhost]
```

eikef commented:
#29123 has been merged instead and while fixing another issue it also solves the same problem as this PR. As such, this PR should be closed.

Thank you mschuett for the archeology required to find this missed fix in any case.

The current devel version should not exhibit this behavior anymore unless my local testing was insufficient. In that case, please ping again :-)

eikef commented:
close_me

angystardust commented:
@eikef unfortunately the current local version still doesn't fix the idempotency in my production scenario


```
---
- name: Create a new host or update an existing host's info
# local_action:
#   module: zabbix_host
  local_action:
    module: zabbix_host
    server_url: https://zabbix.test.corp
    login_user: AnsibleUser
    login_password: password
    host_name: '{{ ansible_fqdn }}'
    visible_name: '{{ ansible_hostname }}'
    host_groups:
      - '{{ zabbix_host_groups | default("My Services") }}'
    link_templates:
      - Template OS Linux
    status: enabled
    state: present
    inventory_mode: disabled
    interfaces:
      - type: 1
        main: 1
        useip: 0
        ip: '{{ ansible_default_ipv4.address }}'
        dns: '{{ ansible_fqdn }}'
        port: 10050
      - type: 1
        main: 0
        useip: 1
        ip: '{{ ansible_default_ipv4.address }}'
        dns: '{{ ansible_fqdn }}'
        port: 10050
    proxy: '{{ zabbix_proxy |default(omit) }}'
```

eikef commented:
The problem here is not the proxy, but the interfaces. Other than use_ip and main they are exactly the same and the logic in the zabbix module to figure out whether they have changed will always return True in that case.
There is not currently any code to fix this in this PR. Once I have bit more time I will see about doing a new issue/PR for this if it is indeed fixable in this case (it ought to be). 

Can you check whether your idempotency returns if you disable one of your two interfaces or change their ports to be distinct from one another?

angystardust commented:
@eikef I've just tried both cases you suggested (one interface or 2 interfaces with different ports) and it works: the idempotency is respected! :)
Any idea why it breaks with my configuration scenario?

eikef commented:
it's somewhere in the following code in zabbix_host:

```
   for exist_interface in exist_interface_list:
            exit_interface_port = int(exist_interface['port'])
            for interface in interfaces:
                interface_port = int(interface['port'])
                if interface_port == exit_interface_port:
                    for key in interface.keys():
                        if str(exist_interface[key]) != str(interface[key]):
                            return True
```

the change detection is not exhaustive enough and it flags a change where there ought to be none. Probably 'main' is not being checked or some such.
